AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/src/HierarchicalVersionUtils.dfy

```diff
@@ -0,0 +1,128 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+include "../Model/AwsCryptographyKeyStoreTypes.dfy"
+include "Structure.dfy"
+include "KMSKeystoreOperations.dfy"
+include "KmsArn.dfy"
+
+module HierarchicalVersionUtils {
+
+  import opened StandardLibrary
+  import opened Wrappers
+  import opened Seq
+
+  import ErrorMessages = KeyStoreErrorMessages
+  import Types = AwsCryptographyKeyStoreTypes
+  import KMS = ComAmazonawsKmsTypes
+  
+  import AtomicPrimitives
+  import UTF8
+  import KMSKeystoreOperations
+  import KmsArn
+  import Structure
+  
+  function method GetMdDigestFromEC(
+    item: Types.EncryptionContextString
+  ) : (output: Types.EncryptionContextString)
+  ensures output.Keys == item.Keys - {Structure.TABLE_FIELD}
+  ensures forall k :: k in output ==> output[k] == item[k]
+  ensures forall k :: k in output ==> k !in {Structure.TABLE_FIELD}
+  {
+    map k <- item.Keys - {Structure.TABLE_FIELD}
+      :: k := item[k]
+  }
+  
+  method GetHV2EC(
+    ecStringMap: Types.EncryptionContextString
+  ) returns (output: Types.EncryptionContextString)
+    ensures Structure.Hv2EncryptionContext?(output)
+  {
+    var withoutRestrictedField := RemoveRestrictedFields(ecStringMap);
+    var items := withoutRestrictedField.Items;
+    var newMap: map<string, string> := map[];
+
+    while items != {}
+        decreases |items|
+    {
+        var item :| item in items;
+        items := items - { item };
+        if (|item.0| >= |Structure.ENCRYPTION_CONTEXT_PREFIX| && item.0[..|Structure.ENCRYPTION_CONTEXT_PREFIX|] == Structure.ENCRYPTION_CONTEXT_PREFIX) {
+          var newKey := item.0[|Structure.ENCRYPTION_CONTEXT_PREFIX|..];
+          newMap := newMap[newKey := item.1];
+        } else {
+          newMap := newMap[item.0 := item.1];
+        }
+    }
```

Comment:
For M1, this OK... but we probably will need to refactor to a more efficient solution.
Dafny maps are not mutable... so these are new allocations (I think).
i.e: each iteration is a new allocation of the map.

There is a action we can use.

Reply:
Added a TODO for this method. I think I might need to change this method into a function so that i can use it in pre and post condition for verification.